dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ThreadCreation.java

```diff
@@ -185,12 +159,90 @@ public void run(Program program) {
                     }
                 }
             }
+            // FIXME: This only allows joining with child threads that were created by this thread.
+            handlePthreadJoins(thread, tid2ComAddrMap);
         }
 
         IdReassignment.newInstance().run(program);
         logger.info("Number of threads (including main): " + program.getThreads().size());
     }
 
+    private void handlePthreadJoins(Thread thread, Map<Expression, Expression> tid2ComAddrMap) {
+        final TypeFactory types = TypeFactory.getInstance();
+        final ExpressionFactory expressions = ExpressionFactory.getInstance();
+        int joinCounter = 0;
+
+        for (DirectFunctionCall call : thread.getEvents(DirectFunctionCall.class)) {
+            final List<Expression> arguments = call.getArguments();
+            switch (call.getCallTarget().getName()) {
+                case "pthread_join", "__pthread_join" -> {
+                    assert arguments.size() == 2;
+                    final Expression tidExpr = arguments.get(0);
+                    // TODO: support return values for threads
+                    // final Expression returnAddr = arguments.get(1);
+
+                    final Register resultRegister = getResultRegister(call);
+                    assert resultRegister.getType() instanceof IntegerType;
+
+                    // We create a switch-case over all possible threads to join with
+                    final Register joinDummyReg = thread.getOrNewRegister("__join#" + joinCounter, types.getBooleanType());
+                    final Label joinEnd = EventFactory.newLabel("L__joinEnd#" + joinCounter);
+
+                    // Construct a case for each possible tid
+                    final Map<Expression, List<Event>> tid2joinCases = new LinkedHashMap<>();
+                    for (Expression tidCandidate : tid2ComAddrMap.keySet()) {
+                        final int tid = ((IValue)tidCandidate).getValueAsInt();
```

Comment:
is this cast safe? If we know values are always `IValue`, why not make the map know about this type?